### DISCUSSION TITLE
How can a RabbitMQ cluster sustain a throughput of 100K messages per second?

### DISCUSSION THREAD

**Discussion by TaheJdidi** [rabbitmq/rabbitmq-dotnet-client#1537] [General]
### Describe the bug

Hi there,

I've created a RabbitMQ client using the rabbitmq.client package. I'm receiving 100,000 messages from a third party every second, and I want to write all of these messages to the broker. I'm sending a batch publish to the broker every second. However, it seems like the broker can't handle this volume of messages every second.

### Reproduction steps

1.Send write per batch every 1s containing 100k messages

### Expected behavior

The broker could write the 100k messages every 1s

**Comment by michaelklishin** [comment]
@TaheJdidi do you really expect that there is a single number that would work for every workload and every environment?

The answer depends on the infrastructure used. A single queue, including quorum queues in 3.13, usually do not reach 100K but several queues collectively can reach hundreds of thousands given enough resources.

In addition to RabbitMQ itself, the answer will vary from client library to client library. A single connection (app instance) that uses Bunny, Pika won't usually go above 10K. Java, .NET, Rust and C++ clients reach tends of thousands.

Throughput will be proportionally lower with larger messages for fairly obvious network and disk I/O reasons.

## Don't Guess

When in doubt, don't guess. Benchmark your actual deployment with multiple queues and streams,
[PerfTest](https://perftest.rabbitmq.com/) and [Stream PerfTest](https://rabbitmq.github.io/rabbitmq-stream-perf-test/stable/htmlsingle/) can simulate a wide range of workloads and use pre-configured topologies.

## Streams

A single stream with a [fast RabbitMQ stream protocol client](https://github.com/rabbitmq/rabbitmq-stream-java-client) can [reach a million](https://www.rabbitmq.com/blog/2021/07/13/rabbitmq-streams-overview). Several streams — a few millions, provided the disks are fast enough. A [superstream](https://www.rabbitmq.com/blog/2022/07/13/rabbitmq-3-11-feature-preview-super-streams) (a partitioned stream) is a way of using multiple streams.

[RabbitMQ blog contains many benchmark results](https://www.rabbitmq.com/blog/tags/performance) that include not only some numbers but also a description of the environment.
